File: flutter_app/lib/presentation/caregiver/job_feed_view_model.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../domain/model/data_models.dart';

class JobFeedUiState {
  final bool isLoading;
  final List<Booking> jobs;
  final String? errorMessage;

  JobFeedUiState({
    this.isLoading = true,
    this.jobs = const [],
    this.errorMessage,
  });

  JobFeedUiState copyWith({
    bool? isLoading,
    List<Booking>? jobs,
    String? errorMessage,
  }) {
    return JobFeedUiState(
      isLoading: isLoading ?? this.isLoading,
      jobs: jobs ?? this.jobs,
      errorMessage: errorMessage ?? this.errorMessage,
    );
  }
}

class JobFeedViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  JobFeedUiState _uiState = JobFeedUiState();
  JobFeedUiState get uiState => _uiState;

  JobFeedViewModel() {
    fetchAvailableJobs();
  }

  void _setUiState(JobFeedUiState state) {
    _uiState = state;
    notifyListeners();
  }

  void fetchAvailableJobs() {
    final userId = _auth.currentUser?.uid;
    if (userId == null) return;

    _firestore.collection("bookings")
        .where("status", whereIn: ["BROADCASTED", "PENDING"])
        .snapshots()
        .listen((snapshot) {
      final jobList = snapshot.docs.map((doc) => Booking.fromMap(doc.data()))
          .where((it) => !it.rejectedBy.contains(userId))
          .where((it) => it.status == "BROADCASTED" || (it.status == "PENDING" && it.caregiverId == userId))
          .where((booking) {
            if (booking.isSubscriptionBooking && booking.allowedCaregivers.isNotEmpty) {
              return booking.allowedCaregivers.contains(userId);
            }
            return true;
          }).toList();

      jobList.sort((a, b) {
        if (a.isEmergency != b.isEmergency) {
          return a.isEmergency ? -1 : 1;
        }
        return b.requestedTime.compareTo(a.requestedTime);
      });

      _setUiState(_uiState.copyWith(isLoading: false, jobs: jobList));
    }, onError: (error) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: error.toString()));
    });
  }

  Future<void> rejectJob(String jobId, Function(bool, String) onResult) async {
    try {
      final userId = _auth.currentUser?.uid;
      if (userId == null) return;
      final bookingRef = _firestore.collection("bookings").doc(jobId);

      final doc = await bookingRef.get();
      if (!doc.exists) return;
      
      final status = doc.data()?['status'] as String?;
      final caregiverId = doc.data()?['caregiverId'] as String?;

      if (status == "PENDING" && caregiverId == userId) {
        await bookingRef.update({"status": "REJECTED"});
      } else {
        await bookingRef.update({
          "rejectedBy": FieldValue.arrayUnion([userId])
        });
      }

      onResult(true, "රැකියාව ප්‍රතික්ෂේප කරන ලදී.");
    } catch (e) {
      onResult(false, e.toString());
    }
  }

  Future<void> acceptJob(String jobId, Function(bool, String) onResult) async {
    try {
      final caregiverId = _auth.currentUser?.uid;
      if (caregiverId == null) return;
      
      final bookingRef = _firestore.collection("bookings").doc(jobId);
      final caregiverDoc = await _firestore.collection("users").doc(caregiverId).get();
      
      final hourlyRate = caregiverDoc.data()?['hourlyRate'] ?? 0.0;
      
      const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
      const digits = "0123456789";
      final rnd = Random();
      final code = "${chars[rnd.nextInt(chars.length)]}${List.generate(4, (_) => digits[rnd.nextInt(digits.length)]).join()}";

      final currentBookingDoc = await bookingRef.get();
      final currentBooking = Booking.fromMap(currentBookingDoc.data()!);

      final Map<String, dynamic> updates = {
        "status": "ACCEPTED",
        "caregiverId": caregiverId,
        "startCode": code,
      };

      if (!currentBooking.isSubscriptionBooking) {
        updates["hourlyRate"] = hourlyRate;
      }

      await bookingRef.update(updates);
      await _firestore.collection("users").doc(caregiverId).update({"isBusy": true});

      // TODO: Schedule job reminders (WorkManager)

      onResult(true, "රැකියාව සාර්ථකව භාර ගන්නා ලදී!");
    } catch (e) {
      onResult(false, e.toString());
    }
  }
}

File: app/src/main/java/com/kina/care/presentation/caregiver/JobFeedViewModel.kt
package com.kina.care.presentation.caregiver

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.workDataOf
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.Booking
import com.kina.care.services.JobReminderWorker
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.Calendar
import java.util.concurrent.TimeUnit

data class JobFeedUiState(
    val isLoading: Boolean = true,
    val jobs: List<Booking> = emptyList(),
    val errorMessage: String? = null
)

class JobFeedViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    private val _uiState = MutableStateFlow(JobFeedUiState())
    val uiState: StateFlow<JobFeedUiState> = _uiState.asStateFlow()

    init { fetchAvailableJobs() }

    private fun fetchAvailableJobs() {
        val userId = auth.currentUser?.uid ?: return

        firestore.collection("bookings")
            .whereIn("status", listOf("BROADCASTED", "PENDING"))
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    _uiState.value = JobFeedUiState(isLoading = false, errorMessage = error.localizedMessage)
                    return@addSnapshotListener
                }

                if (snapshot != null) {
                    val jobList = snapshot.documents.mapNotNull { it.toObject(Booking::class.java)?.copy(id = it.id) }
                        .filter { !it.rejectedBy.contains(userId) }
                        .filter { it.status == "BROADCASTED" || (it.status == "PENDING" && it.caregiverId == userId) }

                        // Premium Package එකක් නම්, Admin Select කරපු අයට විතරයි පෙන්වන්නේ
                        .filter { booking ->
                            if (booking.isSubscriptionBooking && booking.allowedCaregivers.isNotEmpty()) {
                                booking.allowedCaregivers.contains(userId)
                            } else {
                                true // සාමාන්‍ය Public Job එකක්
                            }
                        }

                    val sortedList = jobList.sortedWith(compareByDescending<Booking> { it.isEmergency }.thenByDescending { it.requestedTime })
                    _uiState.value = JobFeedUiState(isLoading = false, jobs = sortedList)
                }
            }
    }

    fun rejectJob(jobId: String, onResult: (Boolean, String) -> Unit) {
        viewModelScope.launch {
            try {
                val userId = auth.currentUser?.uid ?: return@launch
                val bookingRef = firestore.collection("bookings").document(jobId)

                val doc = bookingRef.get().await()
                val status = doc.getString("status")
                val caregiverId = doc.getString("caregiverId")

                // 🌟 අලුත් වෙනස: Direct Request එකක් නම් කෙලින්ම Status එක REJECTED කරනවා
                if (status == "PENDING" && caregiverId == userId) {
                    bookingRef.update("status", "REJECTED").await()
                } else {
                    // Broadcast එකක් නම් මේ Caregiver ට විතරක් නොපෙනී යන්න rejectedBy එකට දානවා
                    bookingRef.update("rejectedBy", FieldValue.arrayUnion(userId)).await()
                }

                onResult(true, "රැකියාව ප්‍රතික්ෂේප කරන ලදී.")
            } catch (e: Exception) {
                onResult(false, e.localizedMessage ?: "Error rejecting job")
            }
        }
    }

    fun acceptJob(context: Context, jobId: String, onResult: (Boolean, String) -> Unit) {
        viewModelScope.launch {
            try {
                val caregiverId = auth.currentUser?.uid ?: return@launch
                val bookingRef = firestore.collection("bookings").document(jobId)

                val caregiverDoc = firestore.collection("users").document(caregiverId).get().await()
                val hourlyRate = caregiverDoc.getDouble("hourlyRate") ?: 0.0

                val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                val digits = "0123456789"
                val code = "${chars.random()}${(1..4).map { digits.random() }.joinToString("")}"

                val updates = mutableMapOf<String, Any>(
                    "status" to "ACCEPTED",
                    "caregiverId" to caregiverId,
                    "startCode" to code
                )

                val currentBooking = bookingRef.get().await().toObject(Booking::class.java)
                if (currentBooking?.isSubscriptionBooking != true) {
                    updates["hourlyRate"] = hourlyRate
                }

                bookingRef.update(updates).await()

                firestore.collection("users").document(caregiverId).update("isBusy", true).await()

                if (currentBooking != null && currentBooking.scheduledTime > 0L) {
                    scheduleJobReminders(context, currentBooking.scheduledTime)
                }

                onResult(true, "රැකියාව සාර්ථකව භාර ගන්නා ලදී!")
            } catch (e: Exception) {
                onResult(false, e.localizedMessage ?: "දෝෂයකි! නැවත උත්සාහ කරන්න.")
            }
        }
    }

    private fun scheduleJobReminders(context: Context, scheduledTimeMs: Long) {
        val workManager = WorkManager.getInstance(context)
        val currentTime = System.currentTimeMillis()

        val oneDayBeforeMs = scheduledTimeMs - (24 * 60 * 60 * 1000L)
        if (oneDayBeforeMs > currentTime) {
            val delay = oneDayBeforeMs - currentTime
            val req1 = OneTimeWorkRequestBuilder<JobReminderWorker>()
                .setInitialDelay(delay, TimeUnit.MILLISECONDS)
                .setInputData(workDataOf("title" to "Upcoming Job Reminder \u23F0", "message" to "හෙට දිනට වෙන්කළ රැකියාවක් ඇත. සූදානම් වන්න.", "type" to "tomorrow"))
                .build()
            workManager.enqueue(req1)
        }

        val cal = Calendar.getInstance().apply { timeInMillis = scheduledTimeMs }
        cal.set(Calendar.HOUR_OF_DAY, 6)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        val morningOfMs = cal.timeInMillis

        if (morningOfMs > currentTime && morningOfMs < scheduledTimeMs) {
            val delay = morningOfMs - currentTime
            val req2 = OneTimeWorkRequestBuilder<JobReminderWorker>()
                .setInitialDelay(delay, TimeUnit.MILLISECONDS)
                .setInputData(workDataOf("title" to "Today is your Job! \uD83D\uDCC5", "message" to "අද දින ඔබ භාරගත් රැකියාව ආරම්භ කිරීමට නියමිතයි. ප්‍රමාද නොවන්න!", "type" to "today"))
                .build()
            workManager.enqueue(req2)
        }
    }
}